src/samplers/cpu/cpi/mod.rs

```diff
@@ -13,71 +15,157 @@ fn init(config: &Config) -> Box<dyn Sampler> {
-            }
+            Ok(counter) => counter,
             Err(_) => {
                 error!("failed to initialize cpu cycles perf counter");
                 return Err(());
             }
-        }
+        };
 
-        match Builder::new()
-            .group(&mut group)
-            .kind(Hardware::INSTRUCTIONS)
-            .build()
+        let instructions = match Builder::new(Hardware::INSTRUCTIONS)
+            .one_cpu(cpu)
+            .any_pid()
+            .exclude_hv(false)
+            .exclude_kernel(false)
+            .build_with_group(&mut cycles)
         {
-            Ok(counter) => {
-                counters.push((Counter::new(&CPU_INSTRUCTIONS, None), counter));
-            }
+            Ok(counter) => counter,
             Err(_) => {
                 error!("failed to initialize cpu instructions perf counter");
                 return Err(());
             }
+        };
+
+        let tsc_event = match Msr::new(MsrId::TSC) {
+            Ok(e) => e,
+            Err(_) => return Err(()),
+        };
+        let tsc = match Builder::new(tsc_event)
+            .one_cpu(cpu)
+            .any_pid()
+            .exclude_hv(false)
+            .exclude_kernel(false)
+            .build_with_group(&mut cycles)
+        {
+            Ok(e) => e,
+            Err(_) => return Err(()),
+        };
+
+        let aperf_event = match Msr::new(MsrId::APERF) {
+            Ok(e) => e,
+            Err(_) => return Err(()),
+        };
+        let aperf = match Builder::new(aperf_event)
+            .one_cpu(cpu)
+            .any_pid()
+            .exclude_hv(false)
+            .exclude_kernel(false)
+            .build_with_group(&mut cycles)
+        {
+            Ok(e) => e,
+            Err(_) => return Err(()),
+        };
+
+        let mperf_event = match Msr::new(MsrId::MPERF) {
+            Ok(e) => e,
+            Err(_) => return Err(()),
+        };
+        let mperf = match Builder::new(mperf_event)
+            .one_cpu(cpu)
+            .any_pid()
+            .exclude_hv(false)
+            .exclude_kernel(false)
+            .build_with_group(&mut cycles)
+        {
+            Ok(e) => e,
+            Err(_) => return Err(()),
+        };
+        // enable the group
+        match cycles.enable_group() {
+            Ok(_) => Ok(Self {
+                cpu,
+                cycles,
+                instructions,
+                tsc,
+                aperf,
+                mperf,
+            }),
+            Err(_) => Err(()),
         }
+    }
+    pub fn read_group(&mut self) -> Result<GroupData, std::io::Error> {
+        return self.cycles.read_group();
+    }
+}
+
+pub struct Cpi {
+    prev: Instant,
+    next: Instant,
+    interval: Duration,
+    groups: Vec<PerfGroup>,
+    // we need to keep the last perf event readings because we need to compute gauge IPC, IPNS, and running CPU frequency.
+    prev_group_data: Option<Vec<GroupData>>,
```

Comment:
@brayniac Is this the right way to track the previous perf counters so that it could compute the IPC for the sampling period?

Reply:
Actually, in-context of the whole sampler, I'm not sure why you're using an Option<> around this when you populate the prev_group_data in the constructor.